sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -848,4 +852,46 @@ object DateTimeUtils {
     val sinceEpoch = BigDecimal(timestamp) / MICROS_PER_SECOND + offset
     new Decimal().set(sinceEpoch, 20, 6)
   }
+
+  def currentTimestamp(): SQLTimestamp = instantToMicros(Instant.now())
+
+  private def today(zoneId: ZoneId): ZonedDateTime = {
+    Instant.now().atZone(zoneId).`with`(LocalTime.MIDNIGHT)
+  }
+
+  private val specialValue = """(EPOCH|NOW|TODAY|TOMORROW|YESTERDAY)\p{Blank}*(.*)""".r
+
+  /**
+   * Converts notational shorthands that are converted to ordinary timestamps.
+   * @param input - a trimmed string
+   * @param zoneId - zone identifier used to get the current date.
+   * @return some of microseconds since the epoch if the conversion completed
+   *         successfully otherwise None.
+   */
+  def convertSpecialTimestamp(input: String, zoneId: ZoneId): Option[SQLTimestamp] = {
+    def isValidZoneId(z: String): Boolean = {
+      z == "" || Try { getZoneId(z) }.isSuccess
+    }
+
+    if (input.length < 3 || !input(0).isLetter) return None
+    input.toUpperCase(Locale.US) match {
+      case specialValue("EPOCH", z) if isValidZoneId(z) => Some(0)
+      case specialValue("NOW", "") => Some(currentTimestamp())
```

Comment:
None if a zone id is defined instead of ignoring it?
```
postgres=# select timestamp 'now CET';
ERROR:  invalid input syntax for type timestamp: "now CET"
LINE 1: select timestamp 'now CET';
```

Reply:
I have already fixed that.